include/ensmallen_bits/age_moea/agemoea_impl.hpp

```diff
@@ -0,0 +1,680 @@
+      else
+      {
+        value = 2.0 * (1.0 - rand) + 2.0 *(rand - 0.5) *
+            std::pow(lowerDelta, distributionIndex + 1.0);
+        perturbationFactor = 1.0 - std::pow(value, mutationPower);
+      }
+
+      candidate(geneIdx) += perturbationFactor * geneRange;
+    }
+    //! Enforce bounds.
+    candidate = arma::min(arma::max(candidate, lowerBound), upperBound);
+}
+
+template <typename MatType>
+void AGEMOEA::NormalizeFront(std::vector<MatType>& population,
+                    arma::colvec& normalization,
+                    const std::vector<size_t>& front,
+                    const arma::Row<size_t>& extreme)
+{
+  arma::Mat<typename MatType::elem_type> vectorizedFront(numVariables, front.size());
+  arma::Col<typename MatType::elem_type> temp;
+  for (size_t i = 0; i< front.size(); i++)
+  {
+    vectorizedFront.col(i) = arma::vectorise(population[front[i]], 0); 
+  }
+  arma::uvec unique = arma::find_unique(extreme);
+  if (extreme.n_elem != unique.n_elem)
+  {
+    normalization = arma::max(vectorizedFront, 1);
+  }
+  arma::colvec hyperplane = arma::solve(vectorizedFront, new arma::colvec(numVariables, arma::fill::ones));
+  if (hyperplane.has_inf() || hyperplane.has_nan() || arma::accu(hyperplane < 0.0) > 0)
+  {
+    normalization = arma::max(vectorizedFront, 1);
+  }
+  normalization = 1. / hyperplane;   
+  if (hyperplane.has_inf() || hyperplane.has_nan())
+  {    
+    normalization = arma::max(vectorizedFront, 1);
+  }
+  normalization = normalization + (normalization == 0);
+  vectorizedFront.each_col( [normalization](arma::vec& a){ a = a / normalization; } );
+  for (size_t i = 0; i< front.size(); i++)
+  {
+    temp = vectorizedFront.col(i);
+    population[front[i]] = temp.reshape(arma::size(population[0]));
+  }
+}
+
+template <typename MatType>
+inline size_t AGEMOEA::GetGeometry(const std::vector<MatType>& population,
+                    const std::vector<size_t>& front,
+                    const std::vector<size_t>& extreme,
+                    size_t m)
+{
+  arma::rowvec d = PointToLineDistance<MatType> (population, front, 
+                                    new arma::rowvec(numVariables, arma::fill::zeros),
+                                    new arma::rowvec(numVariables, arma::fill::ones));
+  for (size_t i = 0; i < extreme.size(); i++)
+  {
+    d[extreme[i]] = arma::datum::inf;
+  }
+  size_t index = arma::index_min(d);
+  double avg = arma::accu(population[front[index]]) / static_cast<double> (numVariables); 
+  double p = std::log(m) / std::log(1.0 / avg);
+  if (p <= 0.1 || std::isnan(p)) {p = 1.0;}
+  return p;
+}
+
+//! Pairwise distance for each point in the given front.
+template <typename MatType>
+inline arma::mat& AGEMOEA::PairwiseDistance(std::vector<MatType>& population,
+                                          const std::vector<size_t>& front,
+                                          size_t dimension)
+{
+  arma::mat final(front.size(),front.size(),arma::fill::zeros);
```

Comment:
This applies to some other lines as well, but ideally we avoid using `arma::mat` and use `MatType` throughout the code. That way, we could run this with half precision i.e. `arma::fmat` as well.